Refactor templates and code duplication

dhoverml opened the pull request:
So I was searching for `bind_static_vararg_method` (which sadly doesn't exist), and ended up reading Godot's template code.  I noticed that there was a lot of duplication and decided to see if I could reduce the duplication.
I've summarized the changes and how I tested it below since template code is often difficult to read without knowing the context.

* Collapse multiple, duplicate, specializations into single specializations (with the help of enable-if).
    For example, this
    ``` C++
    template<typename T> struct MyStruct;
    template<> struct MyStruct<MyType> { /* some specialization */ };
    template<> struct MyStruct<const MyType&> { /* identical specialization as just "MyType"*/ };
    ```
    Becomes
    ``` C++
    template<typename T, typename = void> struct MyStruct;
    template<typename T> struct MyStruct<T, EnableIf_t<types_are_same_v<MyType, RemoveRefPointerConst_t<T>>> {
      /* some specialization */
    };
    ```
* Move existing templates to typedefs.h since one file couldn't include core/typedefs.  Figure it doesn't hurt to move them since it increases visibility
* Add `get_type_name()` to `GetTypeInfo` due to the refactor in typed_array.h.  I think the correct place to put this string is in `PropertyInfo`, but wanted to see what others thought first.
* I'm assuming `GetTypeInfo<T>::get_class_info().class_name` is `T::get_class_static()` for Objects and empty string otherwise.
* `VariantObjectClassChecker` was not collapsed, but instead was generalized to handle const and non-const Ref.  I figured this was necessary since the place its used (`VariantCasterAndValidate`), can pass in any type.
* Add a few new template types to core/typedefs.h that are similar, but not the same, as stl templates:
    * `is_void_v` returns void if the base type is void, instead of exactly void
        Test:
        ``` C++
        print_line("std::is_void_v",
                   std::is_void_v<void>,
                   std::is_void_v<void*>,
                   std::is_void_v<void*&>,
                   std::is_void_v<void**>,
                   std::is_void_v<const void*>,
                   std::is_void_v<const void*&>,
                   std::is_void_v<const void**>,
                   std::is_void_v<const void**&>);
        print_line("is_void_v",
                   is_void_v<void>,
                   is_void_v<void*>,
                   is_void_v<void*&>,
                   is_void_v<void**>,
                   is_void_v<const void*>,
                   is_void_v<const void*&>,
                   is_void_v<const void**>,
                   is_void_v<const void**&>);
        ```
        Output:
        ```
        std::is_void_v true false false false false false false false
        is_void_v true true true true true true true true
        ```
    * `RemoveRefPointerConst_t` removes const, ref, and pointer(s) from a type, leaving you with the base type.
        I tried `std::decay` too, but it returned `false` for every case.
        Test:
        ``` C++
        print_line("std",
                   std::is_same_v<int, std::remove_cv_t<std::remove_reference_t<int>>>,
                   std::is_same_v<int, std::remove_cv_t<std::remove_pointer_t<std::remove_reference_t<int&>>>>,
                   std::is_same_v<int, std::remove_cv_t<std::remove_pointer_t<std::remove_reference_t<int*>>>>,
                   std::is_same_v<int, std::remove_cv_t<std::remove_pointer_t<std::remove_reference_t<int*&>>>>,
                   std::is_same_v<int, std::remove_cv_t<std::remove_pointer_t<std::remove_reference_t<int**>>>>,
                   std::is_same_v<int, std::remove_cv_t<std::remove_pointer_t<std::remove_reference_t<int**&>>>>,
                   std::is_same_v<int, std::remove_cv_t<std::remove_pointer_t<std::remove_reference_t<const int&>>>>,
                   std::is_same_v<int, std::remove_cv_t<std::remove_pointer_t<std::remove_reference_t<const int*>>>>,
                   std::is_same_v<int, std::remove_cv_t<std::remove_pointer_t<std::remove_reference_t<const int*&>>>>,
                   std::is_same_v<int, std::remove_cv_t<std::remove_pointer_t<std::remove_reference_t<const int**>>>>,
                   std::is_same_v<int, std::remove_cv_t<std::remove_pointer_t<std::remove_reference_t<const int**&>>>>);

        print_line("custom",
                   std::is_same_v<int, RemoveRefPointerConst_t<int>>,
                   std::is_same_v<int, RemoveRefPointerConst_t<int&>>,
                   std::is_same_v<int, RemoveRefPointerConst_t<int*>>,
                   std::is_same_v<int, RemoveRefPointerConst_t<int*&>>,
                   std::is_same_v<int, RemoveRefPointerConst_t<int**>>,
                   std::is_same_v<int, RemoveRefPointerConst_t<int**&>>,
                   std::is_same_v<int, RemoveRefPointerConst_t<const int&>>,
                   std::is_same_v<int, RemoveRefPointerConst_t<const int*>>,
                   std::is_same_v<int, RemoveRefPointerConst_t<const int*&>>,
                   std::is_same_v<int, RemoveRefPointerConst_t<const int**>>,
                   std::is_same_v<int, RemoveRefPointerConst_t<const int**&>>);
        ```
        Output:
        ```
        std true true true true false false true true true false false
        custom true true true true true true true true true true true
        ```
    * `static_or_reinterpret_cast` (for lack of a better name, considering it optionally calls `const_cast` too) to simplify otherwise complex expressions.

       For example, this
       ``` C++
       T *ret1 = const_cast<T *>(*reinterpret_cast<T *const *>(p_ptr));
       RefCounted *ret2 = const_cast<RefCounted *>(static_cast<const RefCounted *>(o));
       ```
       Becomes
       ``` C++
       T *ret1 = static_or_reinterpret_cast<T*>(p_ptr);
       RefCounted *ret2 = static_or_reinterpret_cast<RefCounted *>(o);
       ```

      Test code to verify that the correct casts were taking place (I also put printfs in `static_or_reinterpret_cast` so I can see which one is being called)
      ``` C++
      Label* l = memnew(Label);
      Node* n0 = static_or_reinterpret_cast<Node*>(l);
      Node* n1 = static_or_reinterpret_cast<Node*>((const Label*)l);
      Node* n2 = static_or_reinterpret_cast<Node*>((void*)l);
      Node* n3 = static_or_reinterpret_cast<Node*>((const void*)l);
      const Node* n4 = static_or_reinterpret_cast<const Node*>(l);
      const Node* n5 = static_or_reinterpret_cast<const Node*>((void*)l);
      const Node* n6 = static_or_reinterpret_cast<const Node*>((const Label*)l);
      const Node* n7 = static_or_reinterpret_cast<const Node*>((const void*)l);
      ```
      The original reason for this function is that collapsing some duplicate template specializations didn't result in identical code.  In particular, `const_cast` was always called instead of conditionally called.  `struct PtrToArg<T, EnableIf_t<std::is_pointer_v<T>>>` in method_ptrcall.h is a good example of this.
    * `is_ptr_to_const_v` determines if a pointer is a pointer-to-const (like `const int*`).
       This template was needed for `static_or_reinterpret_cast` since using `const_cast` is dependent on whether it is working on a pointer-to-const or not.  This does not return true for ref-to-const.
       Test:
       ``` C++
        print_line("is_ptr_to_const_v",
                   is_ptr_to_const_v<int>,
                   is_ptr_to_const_v<int&>,
                   is_ptr_to_const_v<int*>,
                   is_ptr_to_const_v<int*&>,
                   is_ptr_to_const_v<int**>,
                   is_ptr_to_const_v<int**&>,
                   is_ptr_to_const_v<const int&>,
                   is_ptr_to_const_v<const int*>,
                   is_ptr_to_const_v<const int*&>,
                   is_ptr_to_const_v<const int**>,
                   is_ptr_to_const_v<const int**&>);
       ```
       Output:
       ```
       is_ptr_to_const_v false false false false false false false true true true true
       ```

dhoverml committed:
Refactor templates

Mickeon commented:
Marking myself as off-topic but holy _§&%$_.

AThousandShips commented:
While the general style and usage needs to be evaluated it does appear to provide a very slight but real improvement to build time on Windows, unsure what exactly causes it though

These changes are relatively general and in various areas, so might be easier to review this by splitting it into a few separate PRs with specific focus

dhoverml commented:
> I would say however that I'm not so keen on the static_or_reinterpret_cast, it obfuscates what's going on and I'd prefer explicit casting to not cause any confusion

If I remove it, there will be some templates that will always const_cast even when it's not needed.  I don't think there's any harm, and, it's possible the compiler notices this and optimizes it away.  So it's probably okay to leave the change out. I won't know for sure until I try though.

As for splitting the PR, it would turn into these PRs (optimistically)
* Delete unnecessary forwards to TypedArray (trivial, assuming it still compiles)
* Add get_type_name() to GetPropertyInfo (trivial, though on the surface it adds nothing to pre-existing code)
* Collapse templates with EnableIf_t (still a big change, but more manageable)
    * This one might be able to be further split by which template is being simplified.  The main ones are PtrToArg, TypedArray, GetTypeInfo, and a handful that only exist in binder_common.h.

> At a high-level, simplifying these templates would be very welcome, since they are quite tricky to understand and maintain

Thanks!

> This touches a very large amount of the code responsible for encoding/converting arguments for making ptrcalls to GDExtension, and we don't want to accidentally break binary compatibility.

Agree.  I don't have a way to test GDExtension (thoroughly) unfortunately.

dhoverml commented:
I've managed to split this PR into five smaller PRs.

While splitting this PR...
* I did not `add get_type_name() to GetPropertyInfo` since it seemed kind of hacky.  Instead I added a new template called (for a lack of a better name) `GetNameHelper`, which returns either `T::get_static_class()` or `Variant::get_type_name(GetTypeInfo<T>::VARIANT_TYPE)` depending on type `T`.  It is likely that this could be moved somewhere else (maybe `variant.h`), but I'll wait for what reviewers say when we get to that PR.
* At @AThousandShips 's [request](https://github.com/godotengine/godot/pull/88404#issuecomment-1948657012), I deleted `static_or_reinterpret_cast`.  Please note that this means we'll use `const_cast` even when it's unnecessary in some templates.
* `is_ptr_to_const_v` and `is_void_v` were deleted because `static_or_reinterpret_cast` was deleted.
* any additional deltas made by `static_or_reinterpret_cast` were reverted

The squashed result of the five PRs does not equal this PR due to the changes mentioned above, but it's pretty close.

I'll ask for reviews for one or two PRs at a time since the latter PRs depend on the former.
